File: 01207. Unique Number of Occurrences/solution.go
/**
 * Problem: 1207. Unique Number of Occurrences
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func uniqueOccurrences(numbers []int) bool {
	// Create a map to store the count of each number in the input slice.
	counts := make(map[int]int)

	// Iterate over the numbers slice and count the occurrences of each number.
	for i := 0; i < len(numbers); i++ {
		counts[numbers[i]]++
	}

	// Create a map to track if a particular count value has already been seen.
	exists := make(map[int]bool)

	// Iterate over the counts map to check for duplicate occurrence counts.
	for _, count := range counts {
		// If the count has already been seen, return false as occurrences are not unique.
		if exists[count] {
			return false
		}

		// Mark this count as seen.
		exists[count] = true
	}
	
	// If all occurrence counts are unique, return true.
	return true
}

File: 01207. Unique Number of Occurrences/solution.js
/**
 * Problem: 1207. Unique Number of Occurrences
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Checks if all element frequencies are unique
 *
 * @param {number[]} arr - Input array
 *
 * @returns {boolean} - True if all frequencies are unique
 */
const uniqueOccurrences = (arr) => {
  // Sort the input array in ascending order to group identical elements together
  arr.sort((a, b) => a - b)

  // Initialize an array to store the frequency of each unique element
  const frequencies = []

  // Iterate through the sorted array to count occurrences of each unique element
  for (let i = 0; i < arr.length; i++) {
    // Initialize count for the current element
    let count = 1

    // Count consecutive duplicates of the current element
    while (i + 1 < arr.length && arr[i] === arr[i + 1]) {
      count++
      i++
    }

    // Store the frequency count for the current unique element
    frequencies.push(count)
  }

  // Sort the frequencies array to easily check for duplicate frequencies
  frequencies.sort((a, b) => a - b)

  // Check if any two consecutive frequencies are equal (not unique)
  for (let i = 1; i < frequencies.length; i++) {
    if (frequencies[i] === frequencies[i - 1]) return false
  }

  // All frequencies are unique, return true
  return true
}
